File: backend-node/src/ssh.ts
// SSH runner (mirror of backend-py/app/ssh.py) — ssh2, multi-key auto-detect, timeouts.
// The private key never leaves the backend. Output is returned raw; redaction happens upstream.
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { basename, join } from "node:path";
import { Client, type ConnectConfig } from "ssh2";
import { selectedCaseSource } from "./caseSource";
import { config } from "./config";

const MAX_OUTPUT = 20000;

export interface SSHResult {
  exit_code: number;
  stdout: string;
  stderr: string;
  duration_ms: number;
  truncated: boolean;
}

function cap(s: string): [string, boolean] {
  if (s.length > MAX_OUTPUT) return [s.slice(0, MAX_OUTPUT) + "\n…[output truncated]", true];
  return [s, false];
}

function candidateKeys(primary: string): string[] {
  const out: string[] = [];
  const sandboxKey = join(config.sshKeyDir, "bench_incident_key");
  if (selectedCaseSource() === "sandbox_cases" && existsSync(sandboxKey)) out.push(sandboxKey);
  if (primary && !out.includes(primary)) out.push(primary);

  const looksLikePrivateKey = (name: string) => name.endsWith(".pem") || name.endsWith("_key") || name.startsWith("id_");
  const addCandidateKeys = (dir: string) => {
    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      const p = join(dir, entry.name);
      if (entry.isDirectory()) {
        addCandidateKeys(p);
      } else if (entry.isFile() && looksLikePrivateKey(entry.name) && !entry.name.endsWith(".pub") && !out.includes(p)) {
        out.push(p);
      }
    }
  };

  try {
    addCandidateKeys(config.sshKeyDir);
  } catch {
    /* dir may not exist locally */
  }
  return out.length ? out : [primary];
}

export class SSHRunner {
  private conn: Client | null = null;

  constructor(
    private host: string,
    private username: string = config.sshUsername,
    private port: number = 22,
    private keyPath: string = config.sshKeyPath,
    private connectTimeoutMs: number = config.sshConnectTimeout * 1000,
    private commandTimeoutMs: number = config.sshCommandTimeout * 1000,
  ) {}

  private connectWithKey(keyPath: string): Promise<Client> {
    return new Promise((resolve, reject) => {
      let key: Buffer;
      try {
        key = readFileSync(keyPath);
      } catch (e) {
        return reject(new Error(`cannot read key ${keyPath}: ${(e as Error).message}`));
      }
      const client = new Client();
      const onError = (err: Error) => {
        client.end();
        reject(err);
      };
      client.on("ready", () => {
        client.removeListener("error", onError);
        resolve(client);
      });
      client.on("error", onError);
      const opts: ConnectConfig = {
        host: this.host,
        port: this.port,
        username: this.username,
        privateKey: key,
        readyTimeout: this.connectTimeoutMs,
      };
      client.connect(opts);
    });
  }

  async connect(): Promise<void> {
    if (this.conn) return;
    const errors: string[] = [];
    for (const keyPath of candidateKeys(this.keyPath)) {
      try {
        this.conn = await this.connectWithKey(keyPath);
        this.keyPath = keyPath;
        return;
      } catch (e) {
        const err = e as Error & { level?: string };
        errors.push(`${basename(keyPath)}: ${err.message}`);
        // Only auth failures are worth retrying with another key; bail fast on network errors.
        if (err.level && err.level !== "client-authentication") {
          throw new Error(`SSH connect to ${this.host}:${this.port} failed: ${err.message}`);
        }
      }
    }
    throw new Error(`SSH auth to ${this.host}:${this.port} failed for all keys [${errors.join("; ") || "no keys"}]`);
  }

  async run(command: string, timeoutMs: number = this.commandTimeoutMs): Promise<SSHResult> {
    await this.connect();
    const conn = this.conn!;
    const start = Date.now();
    return new Promise<SSHResult>((resolve, reject) => {
      let out = "";
      let err = "";
      let done = false;
      const finish = (r: SSHResult) => {
        if (done) return;
        done = true;
        clearTimeout(timer);
        resolve(r);
      };
      const timer = setTimeout(() => {
        const [so] = cap(out);
        const [se] = cap(err);
        finish({ exit_code: 124, stdout: so, stderr: se + "\n[command timed out]", duration_ms: Date.now() - start, truncated: true });
      }, timeoutMs);

      conn.exec(command, (e, stream) => {
        if (e) {
          clearTimeout(timer);
          return reject(e);
        }
        stream
          .on("close", (code: number | null) => {
            const [so, t1] = cap(out);
            const [se, t2] = cap(err);
            finish({ exit_code: code ?? 0, stdout: so, stderr: se, duration_ms: Date.now() - start, truncated: t1 || t2 });
          })
          .on("data", (d: Buffer) => {
            out += d.toString();
          });
        stream.stderr.on("data", (d: Buffer) => {
          err += d.toString();
        });
      });
    });
  }

  close(): void {
    if (this.conn) {
      try {
        this.conn.end();
      } catch {
        /* ignore */
      }
      this.conn = null;
    }
  }
}

File: backend-py/app/ssh.py
"""SSH runner — runs ONE approved command on a customer VM, with timeouts.
(rubric E: separate module; rubric C: the runner only executes what the safety layer + human allow.)

The private key never leaves the backend. Output is returned raw; redaction happens at the
audit/activity layer (safety.redact) before anything is stored or sent to the ERP.
"""
from __future__ import annotations

import glob
import os
import socket
import time
from dataclasses import asdict, dataclass

import paramiko
from paramiko.ssh_exception import AuthenticationException, NoValidConnectionsError

from .case_source import selected_case_source
from .config import settings

_MAX_OUTPUT = 20000  # cap stdout/stderr to keep audit + prompts bounded


@dataclass
class SSHResult:
    exit_code: int
    stdout: str
    stderr: str
    duration_ms: int
    truncated: bool

    def as_dict(self) -> dict:
        return asdict(self)


class SSHError(Exception):
    pass


def _load_key(path: str):
    # paramiko 3.2+ autodetects key type; fall back to trying each type.
    try:
        return paramiko.PKey.from_path(path)  # type: ignore[attr-defined]
    except Exception:
        pass
    for cls in (paramiko.Ed25519Key, paramiko.ECDSAKey, paramiko.RSAKey, paramiko.DSSKey):
        try:
            return cls.from_private_key_file(path)
        except Exception:
            continue
    raise SSHError(f"Could not load SSH private key at {path}")


def _cap(raw: bytes) -> tuple[str, bool]:
    text = raw.decode("utf-8", errors="replace")
    if len(text) > _MAX_OUTPUT:
        return text[:_MAX_OUTPUT] + "\n…[output truncated]", True
    return text, False


class SSHRunner:
    """One connection per troubleshooting run; reused across that run's commands."""

    def __init__(
        self,
        host: str,
        username: str | None = None,
        port: int = 22,
        key_path: str | None = None,
        connect_timeout: int | None = None,
        command_timeout: int | None = None,
    ) -> None:
        self.host = host
        self.username = username or settings.ssh_username
        self.port = port or 22
        self.key_path = key_path or settings.ssh_private_key_path
        self.connect_timeout = connect_timeout or settings.ssh_connect_timeout
        self.command_timeout = command_timeout or settings.ssh_command_timeout
        self._client: paramiko.SSHClient | None = None

    def _candidate_keys(self) -> list[str]:
        # Try the sandbox bench key first when local Docker cases are active, then the
        # configured key, then the real per-case *.pem keys. This avoids switching env files
        # between rehearsal VMs and local fake VMs.
        cands: list[str] = []
        if selected_case_source() == "sandbox_cases":
            sandbox_key = os.path.join(settings.ssh_key_dir, "bench_incident_key")
            if os.path.exists(sandbox_key):
                cands.append(sandbox_key)
        if self.key_path:
            if self.key_path not in cands:
                cands.append(self.key_path)
        try:
            patterns = ("**/*.pem", "**/*_key", "**/id_*")
            for pattern in patterns:
                for p in sorted(glob.glob(os.path.join(settings.ssh_key_dir, pattern), recursive=True)):
                    if p.endswith(".pub") or not os.path.isfile(p):
                        continue
                    if p not in cands:
                        cands.append(p)
        except Exception:
            pass
        return cands or [self.key_path]

    def connect(self) -> None:
        if self._client is not None:
            return
        errors: list[str] = []
        for key_path in self._candidate_keys():
            client = paramiko.SSHClient()
            client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
            try:
                client.connect(
                    hostname=self.host,
                    port=self.port,
                    username=self.username,
                    pkey=_load_key(key_path),
                    timeout=self.connect_timeout,
                    banner_timeout=self.connect_timeout,
                    auth_timeout=self.connect_timeout,
                    look_for_keys=False,
                    allow_agent=False,
                )
            except (socket.timeout, NoValidConnectionsError, OSError) as exc:
                # Network/host problem — no point trying other keys.
                self._safe_close(client)
                raise SSHError(f"SSH connect to {self.host}:{self.port} failed: {exc}") from exc
            except (AuthenticationException, SSHError) as exc:
                errors.append(f"{os.path.basename(key_path)}: {exc}")
                self._safe_close(client)
                continue
            except Exception as exc:
                errors.append(f"{os.path.basename(key_path)}: {exc}")
                self._safe_close(client)
                continue
            self._client = client
            self.key_path = key_path
            return
        raise SSHError(f"SSH auth to {self.host}:{self.port} failed for all keys [{'; '.join(errors) or 'no keys found'}]")

    @staticmethod
    def _safe_close(client: "paramiko.SSHClient") -> None:
        try:
            client.close()
        except Exception:
            pass

    def run(self, command: str, timeout: int | None = None) -> SSHResult:
        self.connect()
        assert self._client is not None
        timeout = timeout or self.command_timeout
        start = time.monotonic()
        timed_out = False
        try:
            stdin, stdout, stderr = self._client.exec_command(command, timeout=timeout)
            stdout.channel.settimeout(timeout)
            out_b = stdout.read()
            err_b = stderr.read()
            code = stdout.channel.recv_exit_status()
        except socket.timeout:
            timed_out = True
            code, out_b, err_b = 124, b"", b"[command timed out]"
        except Exception as exc:
            raise SSHError(f"command execution failed on {self.host}: {exc}") from exc
        out, t1 = _cap(out_b)
        err, t2 = _cap(err_b)
        return SSHResult(
            exit_code=code,
            stdout=out,
            stderr=err,
            duration_ms=int((time.monotonic() - start) * 1000),
            truncated=timed_out or t1 or t2,
        )

    def close(self) -> None:
        if self._client is not None:
            try:
                self._client.close()
            finally:
                self._client = None

    def __enter__(self) -> "SSHRunner":
        self.connect()
        return self

    def __exit__(self, *exc) -> None:
        self.close()
